### PR TITLE
Callback api for maxmind-db-reader

### PR DESCRIPTION
**Pull request by erik-humio**
This introduces an alternative API for looking up records without allocating objects on the heap.
(The normal API allocates ~6KB per lookup when using DatabaseReader, and ~1.7KB per lookup when using Reader.)
With a callback-style API, this can be avoided if different trade-offs are desired.

### CODE REVIEW

File: src/main/java/com/maxmind/db/CallbackDecoder.java [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+						   + ")");
+            }
+
+            type = Type.get(typeNum);
+        }
+
+        int size = ctrlByte & 0x1f;
+        if (size >= 29) {
+            switch (size) {
+	    case 29:
```
Comment (mpdn) [inline src/main/java/com/maxmind/db/CallbackDecoder.java:119]: Nit: Don't mix tabs and spaces. Maybe just autoformat the code in intellij.

**Review by mpdn — COMMENTED** [review]
This looks good and seems like a definite improvement over the current reader as so far as making fewer allocations. I'm not familiar with the internal structure of maxmind DBs, so I don't feel like I can provide a really comprehensive review.

I do have one nagging feeling though - and that is whether it is really necessary to manage an on-disk database at all. Maxmind Geolite2 City IP db is 64MB decompressed - we could easily just load that into a data structure once and be done with it (although the Maxmind reader does not support iteration). Do we know we need to eg. be compatible with other, larger databases or something like that?

I hope it's not out of line for me to suggest this - you have clearly put a lot of effort into this.

**Comment by erik-humio** [comment]
@mpdn Can we merge this?

**Comment by mpdn** [comment]
@erik-humio Sure, if you've read and considered my last comment, feel free to merge.

**Comment by pmech** [comment]
Is this ready to merge? If so, can you do that @erik-humio .

**Comment by pmech** [comment]
@erik-humio it is very strange. I can normally assign people to issues/pr's, but for this I cannot. It is like I only have read access to this issue. Very strange.

What is the status here?

**Comment by pmech** [comment]
Status? @erik-humio

**Comment by pmech** [comment]
Status?

**Pull request merged**